uportal-war/src/main/java/org/apereo/portal/groups/GroupsCacheAuthenticationListener.java

```diff
@@ -0,0 +1,95 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apereo.portal.groups;
+
+import java.util.Set;
+
+import net.sf.ehcache.Cache;
+import net.sf.ehcache.Element;
+import org.apereo.portal.EntityIdentifier;
+import org.apereo.portal.security.IPerson;
+import org.apereo.portal.services.IAuthenticationListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+/**
+ * Responsible for removing some membership-related cache entries when a user
+ * authenticates so they can be evaluated afresh.
+ *
+ * @author drewwills
+ */
+@Component
+public class GroupsCacheAuthenticationListener implements IAuthenticationListener {
+
+    private Cache parentGroupsCache;
+
+    private Cache childrenCache;
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Autowired
+    @Qualifier(value = "org.apereo.portal.groups.GroupMemberImpl.parentGroups")
+    public void setParentGroupsCache(Cache parentGroupsCache) {
+        this.parentGroupsCache = parentGroupsCache;
+    }
+
+    @Autowired
+    @Qualifier(value = "org.apereo.portal.groups.EntityGroupImpl.children")
+    public void setChildrenCache(Cache childrenCache) {
+        this.childrenCache = childrenCache;
+    }
+
+    @Override
+    public void userAuthenticated(IPerson user) {
+
+        /*
+         * Used to log the time it takes to complete this operation;  the author
+         * has some anxiety about running time with large numbers of elements in
+         * the cache.
+         */
+        final long timestamp = System.currentTimeMillis();
+
+        /*
+         * Group/member relationships are cached 2 ways:  child-to-parents and
+         * parent-to-children.  We need to flush both.
+         */
+        final EntityIdentifier ei = user.getEntityIdentifier();
+        final Element parentGroupsElement = parentGroupsCache.get(ei);
+        if (parentGroupsElement != null) {
+            // We have some flushing work to do...
+            int numPurged = 1;
+            final Set<IEntityGroup> parentGroups = (Set<IEntityGroup>) parentGroupsElement.getObjectValue();
+            for (IEntityGroup group : parentGroups) {
+                final EntityIdentifier uei = group.getUnderlyingEntityIdentifier();
+                if (childrenCache.remove(uei)) {
+                    ++numPurged;
+                }
+            }
+            parentGroupsCache.remove(ei);
+            logger.debug("Purged {} local group cache entries for authenticated user '{}' in {}ms",
+                    numPurged, user.getUserName(), Long.toBinaryString(System.currentTimeMillis() - timestamp));
```

Comment:
Suggestion: collect the names of the groups purged from and log that. Not much more expensive; more valuable for debugging.